✨ Feature : 공통 예외 핸들러 추가, 로그인 API, 필터를 이용한 토큰 검증 로직 개발

newnyee opened the pull request:
## 💻 작업 내역 
### ✓ 공통 예외 핸들러 추가 9577414bd82b28f953484498a90ca553b242ce1b

### ✓ 로그인 API 개발
- **사용자 인증** a8c7a678ca1e6e0e17dca759feceba4ebf2c2ecf
- **인증 후 토큰 발급** e1dff5f67614931c7aafa2e423c9f356c7270524

### ✓ 필터를 이용한 토큰 검증 로직 개발 e1dff5f67614931c7aafa2e423c9f356c7270524

### ✓ Feed API 명세서 (swagger)작성


## 🔎 PR 특이 사항
- swagger에 보여질 항목에 대한 명칭 정리가 필요할 것 같습니다.
![image](https://github.com/wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service/assets/121937711/6b0be944-4bc7-4b01-988a-c6cf4cfc5053)

- 저번 회의 때 나온 사항으로 공통 예외 핸들러 리턴값(ResponseEntity) body에 들어갈 예외 코드 리턴값(현재 "BAD_REQUEST"로 되어있음) 한번 더 의견 나눠야할 것 같습니다.

newnyee committed:
Refactor: 예외 Handler 코드 추가

newnyee committed:
Refactor: Gradle 코드 추가

newnyee committed:
Feature: UserRepository 메소드 추가

newnyee committed:
Merge branch 'develop' of https://github.com/wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service into feature/FEED-13

# Conflicts:
#	build.gradle
#	src/main/java/com/wanted/feed/user/domain/UserRepository.java

newnyee committed:
Feature: 예외 추가

newnyee committed:
Refactor: SecurityConfig 메소드 변경, 추가

newnyee committed:
Feature: 로그인 토큰인증 구현

newnyee committed:
Feature: 에러 추가

newnyee committed:
Refactor: 사용하지 않는 Exception 제거

newnyee committed:
Refactor: 로그인 API 수정

newnyee committed:
Refactor: 토큰 검증 수정

newnyee committed:
Merge branch 'develop' of https://github.com/wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service into feature/FEED-13

newnyee committed:
Feature: 예외 추가

hsjkdss228 commented on src/main/java/com/wanted/feed/common/JwtFilter.java:
```diff
@@ -0,0 +1,87 @@
+        filterChain.doFilter(request, response);
+    }
+
+    private void jwtExceptionHandler(HttpServletResponse response, Exception ex) {
```
'jwt 예외를 핸들링한다'는 동사형 의미의 `handleJwtException`이라는 명칭을 제안드립니다.

newnyee replied:
말씀하신대로 `handleJwtException` 으로 메서드명 수정하였습니다!

hsjkdss228 commented on src/main/java/com/wanted/feed/common/config/swagger/SecurityConfig.java:
Swagger와 별개의 Security에 대한 config라면 패키지를 분리하는 게 좋을 것 같다는 의견 드립니다!

newnyee replied:
말씀하신대로 config.security 패키지로 분리하였습니다!

newnyee committed:
Refactor: 리뷰 받은 코드 수정

- JwtFilter.jwtExceptionHandler -> JwtFilter.handleJwtException 리네임
- SecurityConfig 클래스 swagger 패키지에서 분리

min050410 commented on src/main/java/com/wanted/feed/common/JwtFilter.java:
```diff
@@ -0,0 +1,87 @@
+
+        final String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
+
+        if(authorization == null) {
```
안녕하세요, 코드리뷰 시작하겠습니다~

if ( 
컨벤션 수정 부탁드립니다!

newnyee replied:
if문 사용시 한칸 띄우는것도 컨벤션이였군요! 말씀대로 수정하였습니다!

min050410 commented on src/main/java/com/wanted/feed/common/JwtFilter.java:
```diff
@@ -0,0 +1,87 @@
+            log.error("[NullTokenException] ex");
+            handleJwtException(response, new NullTokenException());
+            return;
+        } else if (!authorization.startsWith("Bearer ")) {
+            log.error("[InvalidTypeOfTokenException] ex");
+            handleJwtException(response, new InvalidTypeOfTokenException());
+            return;
+        }
```
else if를 쓰지 않아도 될 것 같습니다!
앞에서 이미 return을 하고 있어서요!

newnyee replied:
if문 두개로 분리하여 변경하였습니다!

min050410 commented on src/main/java/com/wanted/feed/common/JwtFilter.java:
```diff
@@ -0,0 +1,87 @@
+            log.error("[NullTokenException] ex");
+            handleJwtException(response, new NullTokenException());
+            return;
+        } else if (!authorization.startsWith("Bearer ")) {
```
이 부분은 매직넘버를 제거하면 좋을 것 같습니당

newnyee replied:
말씀하신대로 상수로 선언하여 코드 수정하였습니다!

min050410 commented on src/main/java/com/wanted/feed/exception/GlobalControllerAdvice.java:
```diff
@@ -14,11 +14,48 @@
 
-}
\ No newline at end of file
+    private String getMessage(MethodArgumentNotValidException e) {
+        StringBuilder stringBuilder = new StringBuilder();
+        boolean isFirst = true;
+
+        for (FieldError fieldError : e.getBindingResult().getFieldErrors()) {
+            if (!isFirst) {
+                stringBuilder.append(", ");
+            } else {
+                isFirst = false;
+            }
+
+            stringBuilder.append("[");
+            stringBuilder.append(fieldError.getField());
+            stringBuilder.append("]은(는) ");
+            stringBuilder.append(fieldError.getDefaultMessage());
+        }
+        return stringBuilder.toString();
+    }
+}
```
isFirst 변수 대신 이런 방법이 있습니다.
```java
StringJoiner joiner = new StringJoiner(", ");

for (FieldError fieldError : e.getBindingResult().getFieldErrors()) {
    String message = "[" + fieldError.getField() + "]은(는) " + fieldError.getDefaultMessage();
    joiner.add(message);
}

return joiner.toString();
```

newnyee replied:
오  코드가 확 주는군요. 말씀하신대로 코드 수정하였습니다! 감사합니다😊

min050410 commented on src/main/java/com/wanted/feed/user/service/LoginService.java:
```diff
@@ -0,0 +1,34 @@
+    @Value("${jwt.expired-ms}")
+    public Long expiredMs;
+
+    public User getAuthenticatedByLogin (LoginRequestDto loginRequestDto) {
+        return userRepository.findByUsernameAndPassword(loginRequestDto.getUsername(),
+                loginRequestDto.getPassword()).orElseThrow(NotFoundUserException::new);
+    }
```
외부에서 해당메서드를 사용하나요?
private으로 수정하는게 더 좋아보입니다!

newnyee replied:
말씀하신대로 private로 수정하였습니다!

min050410 commented on src/main/java/com/wanted/feed/user/service/LoginService.java:
```diff
@@ -0,0 +1,34 @@
+                loginRequestDto.getPassword()).orElseThrow(NotFoundUserException::new);
+    }
+
+    @Transactional
+    public String getLoginAuthorization (LoginRequestDto loginRequestDto) {
+        return TokenProvider.createJwt(getAuthenticatedByLogin(loginRequestDto).getId(), secretKey, expiredMs);
+    }
+}
```
- `@Transactional` 보단 여기서는 `@Transactional(readOnly = true)`가 더 좋을 것 같네요
[https://hungseong.tistory.com/74](https://hungseong.tistory.com/74)
참고 부탁드립니다

- 그리고 getAuthenticatedByLogin메서드 위에 트랜젝션 어노테이션을 붙이는게 맞는 것 같습니다
트랜젝션 작업을 하는 것은 getAuthenticatedByLogin 메서드니까요!

newnyee replied:
참고자료 정말 감사드립니다😊
말씀하신 부분을 수정하는 도중 getAuthenticationByLogin 메서드에 다른 코멘트 달아주신 부분 (public을 private로 변경)을 변경하던 중 에러가 발생하여 이유를 검색해보니,  @Transactioal 어노테이션을 적용하려면 프록시가 생성되어야하는데 그러려면 메소드가 public 상태여야 한다고 합니다!
해서 public으로 다시 변경하여 위의 코멘트 반영하였습니다!

min050410 commented on src/main/java/com/wanted/feed/user/dto/LoginRequestDto.java:
```diff
@@ -0,0 +1,23 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
+@Schema(name = "LoginRequestDto", title = "Login Request", description = "로그인 인증 요청 Object")
+public class LoginRequestDto {
+
+    @Schema(title = "user name", description = "사용자 계정")
+    @NotBlank(message = "필수값 입니다")
+    @Size(min = 2, message = "2글자 이상 입력해주세요")
+    private String username;
+
+    @Schema(title = "password", description = "사용자 비밀번호")
+    @NotBlank(message = "필수값 입니다")
+    @Size(min = 10, message = "10글자 이상 입력해주세요")
+    private String password;
+}
```
LoginRequest에 대한 Validation을 상세하게 작성하니 프론트 친화적이어서 굉장히 좋네요!! 😁

newnyee replied:
감사합니다😊

min050410 reviewed (commented):
코드리뷰 마치겠습니다 😃

마지막으로 인우님이 언급해주신 EOF 스타일 관련 수정해주세요!

kmss6905 commented on src/main/java/com/wanted/feed/common/config/security/SecurityConfig.java:
```diff
@@ -0,0 +1,58 @@
+    private final UserRepository userRepository;
+
+    @Value("${jwt.secret}")
+    private String secretKey;
```
외부 설정 파일을 .gitignore 에 등록했기 때문에 저희가 앱을 실행할 때는 각자의 설정파일에 jwt.secret 설정 값을 넣어야 실행될 것 같습니다. 

@hsjkdss228 @min050410 @buhee1029 @newnyee

@Value 를 사용해서 외부 설정값을 가져오는 경우 개발 단계에 어떻게 적용해야 좋을 지 서로 이야기 해보는 것도 좋을 것 같습니다.
(가령 `application-local.yml` 과 `application.yml` 를 두어 개발 한다던지?)

newnyee replied:
코멘트 감사합니다 민식님! 말씀을 토대로 회의때 이야기 나누겠습니다!

newnyee committed:
Refactor: 리뷰 받은 코드 수정

- if문 컨벤션 수정
- 불필요한 else if문 수정
- 매직넘버 제거
- StringBuilder대신 StringJoiner 사용하여 코드 개선
- 잘못된 어노테이션 위치 변경 (LoginService) & @Transactional 어노테이션에 (readOnly = true) 적용

newnyee committed:
Merge branch 'develop' of https://github.com/wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service into feature/FEED-13

# Conflicts:
#	build.gradle
#	src/main/java/com/wanted/feed/exception/ErrorType.java
#	src/main/java/com/wanted/feed/exception/GlobalControllerAdvice.java

newnyee committed:
Refactor: 에러타입 추가, JwtFilter.handlerJwtException 파라미터 통일

newnyee committed:
Refactor: JWTResponse 추가

토큰 발급 응답 포멧을 위해 JWTResponse 추가

newnyee committed:
Test: Login 관련 test code 추가

토큰 발급, 토큰 유효성검사 테스트 추가
